File: android/core/network/src/main/java/org/circle13/antara/core/network/NoiseHandshake.kt
package org.circle13.antara.core.network

import java.security.SecureRandom
import java.security.KeyPairGenerator
import java.security.KeyFactory
import java.security.spec.X509EncodedKeySpec
import java.security.spec.PKCS8EncodedKeySpec
import javax.crypto.KeyAgreement
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class NoiseHandshake {

    data class KeyPair(val publicKey: ByteArray, val privateKey: ByteArray)
    data class DerivedKeys(val rxKey: ByteArray, val txKey: ByteArray)

    private val secureRandom = SecureRandom()

    fun generateKeyPair(): KeyPair {
        val keyGen = KeyPairGenerator.getInstance("EC")
        keyGen.initialize(256, secureRandom)
        val pair = keyGen.generateKeyPair()
        return KeyPair(pair.public.encoded, pair.private.encoded)
    }

    fun computeDH(localPrivateBytes: ByteArray, remotePublicBytes: ByteArray): ByteArray {
        val keyFactory = KeyFactory.getInstance("EC")
        val privateKey = keyFactory.generatePrivate(PKCS8EncodedKeySpec(localPrivateBytes))
        val publicKey = keyFactory.generatePublic(X509EncodedKeySpec(remotePublicBytes))
        
        val keyAgreement = KeyAgreement.getInstance("ECDH")
        keyAgreement.init(privateKey)
        keyAgreement.doPhase(publicKey, true)
        
        return keyAgreement.generateSecret()
    }

    fun hkdfDerive(sharedSecret: ByteArray, salt: ByteArray): DerivedKeys {
        val prk = hmacSha256(salt, sharedSecret)
        
        // Derive rx and tx keys
        val infoRx = "antara-rx".toByteArray()
        val infoTx = "antara-tx".toByteArray()
        
        val rxKey = hmacSha256(prk, infoRx + 0x01.toByte())
        val txKey = hmacSha256(prk, infoTx + 0x01.toByte())
        
        return DerivedKeys(rxKey, txKey)
    }

    private fun hmacSha256(key: ByteArray, data: ByteArray): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        val secretKey = SecretKeySpec(key, "HmacSHA256")
        mac.init(secretKey)
        return mac.doFinal(data)
    }
}

File: ios/AntaraPackage/Sources/AntaraCrypto/NoiseHandshake.swift
import Foundation
import CryptoKit

public final class NoiseHandshake {
    
    public struct KeyAgreementPair {
        public let privateKey: Curve25519.KeyAgreement.PrivateKey
        public let publicKeyData: Data
        
        public init() {
            let key = Curve25519.KeyAgreement.PrivateKey()
            self.privateKey = key
            self.publicKeyData = key.publicKey.rawRepresentation
        }
    }
    
    public struct DerivedSymmetricKeys {
        public let rxKey: SymmetricKey
        public let txKey: SymmetricKey
        
        public init(rxKey: SymmetricKey, txKey: SymmetricKey) {
            self.rxKey = rxKey
            self.txKey = txKey
        }
    }
    
    public init() {}
    
    public func generateKeyPair() -> KeyAgreementPair {
        return KeyAgreementPair()
    }
    
    public func computeSharedSecret(localPrivate: Curve25519.KeyAgreement.PrivateKey, remotePublicRepresentation: Data) throws -> SharedSecret {
        let remotePublic = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: remotePublicRepresentation)
        return try localPrivate.sharedSecretFromKeyAgreement(with: remotePublic)
    }
    
    public func hkdfDerive(sharedSecret: SharedSecret, salt: Data) -> DerivedSymmetricKeys {
        // Derive rx and tx keys
        let infoRx = "antara-rx".data(using: .utf8)!
        let infoTx = "antara-tx".data(using: .utf8)!
        
        let rxKey = sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: salt,
            sharedInfo: infoRx,
            outputByteCount: 32
        )
        
        let txKey = sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: salt,
            sharedInfo: infoTx,
            outputByteCount: 32
        )
        
        return DerivedSymmetricKeys(rxKey: rxKey, txKey: txKey)
    }
}
